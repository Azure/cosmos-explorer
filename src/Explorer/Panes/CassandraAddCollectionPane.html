<div data-bind="visible: visible, event: { keydown: onPaneKeyDown }">
  <div
    class="contextual-pane-out"
    data-bind="
    click: cancel,
    clickBubble: false"
  ></div>
  <div class="contextual-pane" id="cassandraaddcollectionpane">
    <!-- Add Cassandra collection form - Start -->
    <div class="contextual-pane-in">
      <form
        class="paneContentContainer"
        role="dialog"
        aria-label="Add Table"
        data-bind="
            submit: submit"
      >
        <!-- Add Cassandra collection header - Start -->
        <div class="firstdivbg headerline">
          <span data-bind="text: title"></span>
          <div
            class="closeImg"
            role="button"
            aria-label="Close pane"
            tabindex="0"
            data-bind="
                    click: cancel, event: { keypress: onCloseKeyPress }"
          >
            <img src="../../../images/close-black.svg" title="Close" alt="Close" />
          </div>
        </div>
        <!-- Add Cassandra collection header - End -->
        <!-- Add Cassandra collection errors - Start -->
        <div
          class="warningErrorContainer"
          aria-live="assertive"
          data-bind="visible: formErrors() && formErrors() !== ''"
        >
          <div class="warningErrorContent">
            <span><img class="paneErrorIcon" src="/error_red.svg" alt="Error" /></span>
            <span class="warningErrorDetailsLinkContainer">
              <span class="formErrors" data-bind="text: formErrors, attr: { title: formErrors }"></span>
            </span>
          </div>
        </div>
        <!-- Add Cassandra collection errors - End -->
        <div class="paneMainContent">
          <div class="seconddivpadding">
            <p>
              <span class="mandatoryStar">*</span> Keyspace name
              <span class="infoTooltip" role="tooltip" tabindex="0">
                <img class="infoImg" src="/info-bubble.svg" alt="More information" />
                <span class="tooltiptext infoTooltipWidth"
                  >Select an existing keyspace or enter a new keyspace id.</span
                >
              </span>
            </p>

            <div class="createNewDatabaseOrUseExisting">
              <input
                class="createNewDatabaseOrUseExistingRadio"
                aria-label="Create new keyspace"
                name="databaseType"
                type="radio"
                role="radio"
                id="keyspaceCreateNew"
                data-test="addCollection-newDatabase"
                tabindex="0"
                data-bind="checked: keyspaceCreateNew, checkedValue: true, attr: { 'aria-checked': keyspaceCreateNew() ? 'true' : 'false' }"
              />
              <span class="createNewDatabaseOrUseExistingSpace" for="keyspaceCreateNew">Create new</span>

              <input
                class="createNewDatabaseOrUseExistingRadio"
                aria-label="Use existing keyspace"
                name="databaseType"
                type="radio"
                role="radio"
                id="keyspaceUseExisting"
                data-test="addCollection-existingDatabase"
                tabindex="0"
                data-bind="checked: keyspaceCreateNew, checkedValue: false, attr: { 'aria-checked': !keyspaceCreateNew() ? 'true' : 'false' }"
              />
              <span class="createNewDatabaseOrUseExistingSpace" for="keyspaceUseExisting">Use existing</span>
            </div>

            <input
              id="keyspace-id"
              data-test="addCollection-keyspaceId"
              type="text"
              autocomplete="off"
              pattern="[^/?#\\]*[^/?# \\]"
              title="May not end with space nor contain characters '\' '/' '#' '?'"
              placeholder="Type a new keyspace id"
              size="40"
              class="collid"
              data-bind="visible: keyspaceCreateNew, textInput: keyspaceId, hasFocus: firstFieldHasFocus"
              aria-label="Keyspace id"
              aria-required="true"
              autofocus
            />

            <input
              type="text"
              aria-required="true"
              autocomplete="off"
              pattern="[^/?#\\]*[^/?# \\]"
              title="May not end with space nor contain characters '\' '/' '#' '?'"
              list="keyspacesList"
              placeholder="Choose existing keyspace id"
              size="40"
              class="collid"
              data-bind="visible: !keyspaceCreateNew(), textInput: keyspaceId, hasFocus: firstFieldHasFocus"
              aria-label="Keyspace id"
            />

            <datalist id="keyspacesList" data-bind="foreach: container.nonSystemDatabases">
              <option data-bind="value: $data.id"> </option>
            </datalist>

            <div
              class="databaseProvision"
              aria-label="New database provision support"
              data-bind="visible: keyspaceCreateNew"
            >
              <input
                tabindex="0"
                type="checkbox"
                id="keyspaceSharedThroughput"
                title="Provision shared throughput"
                data-bind="checked: keyspaceHasSharedOffer"
              />
              <span class="databaseProvisionText" for="keyspaceSharedThroughput">Provision keyspace throughput</span>
              <span class="infoTooltip" role="tooltip" tabindex="0">
                <img class="infoImg" src="/info-bubble.svg" alt="More information" />
                <span class="tooltiptext provisionDatabaseThroughput"
                  >Provisioned throughput at the keyspace level will be shared across unlimited number of tables within
                  the keyspace</span
                >
              </span>
            </div>
            <!-- 1 -->
            <!-- ko if: hasAutoPilotV2FeatureFlag && !hasAutoPilotV2FeatureFlag() -->
            <div data-bind="visible: keyspaceCreateNew() && keyspaceHasSharedOffer()">
              <throughput-input-autopilot-v3
                params="{ 
                                    testId: 'cassandraThroughputValue-v3-shared',
                                    value: keyspaceThroughput, 
                                    minimum: minThroughputRU, 
                                    maximum: maxThroughputRU, 
                                    isEnabled: keyspaceCreateNew() && keyspaceHasSharedOffer(), 
                                    label: sharedThroughputRangeText,
                                    ariaLabel: sharedThroughputRangeText,
                                    requestUnitsUsageCost: requestUnitsUsageCostShared,
                                    spendAckChecked: sharedThroughputSpendAck,
                                    spendAckId: 'sharedThroughputSpendAck-v3-shared',
                                    spendAckText: sharedThroughputSpendAckText,
                                    spendAckVisible: sharedThroughputSpendAckVisible,
                                    showAsMandatory: true,
                                    infoBubbleText: ruToolTipText,
                                    throughputAutoPilotRadioId: 'newKeyspace-databaseThroughput-autoPilotRadio-v3-shared',
                                    throughputProvisionedRadioId: 'newKeyspace-databaseThroughput-manualRadio-v3-shared',
                                    isAutoPilotSelected: isSharedAutoPilotSelected,
                                    maxAutoPilotThroughputSet: sharedAutoPilotThroughput,
                                    autoPilotUsageCost: autoPilotUsageCost,
                                    canExceedMaximumValue: canExceedMaximumValue,
                            costsVisible: costsVisible,
                                    showAutoPilot: !isFreeTierAccount()
                                    }"
              >
              </throughput-input-autopilot-v3>
            </div>
            <!-- /ko -->
            <!-- ko if: hasAutoPilotV2FeatureFlag && hasAutoPilotV2FeatureFlag() -->
            <div data-bind="visible: keyspaceCreateNew() && keyspaceHasSharedOffer()">
              <throughput-input
                params="{ 
                                    testId: 'cassandraThroughputValue-v2-shared',
                                    value: keyspaceThroughput, 
                                    minimum: minThroughputRU, 
                                    maximum: maxThroughputRU, 
                                    isEnabled: keyspaceCreateNew() && keyspaceHasSharedOffer(), 
                                    label: sharedThroughputRangeText,
                                    ariaLabel: sharedThroughputRangeText,
                                    requestUnitsUsageCost: requestUnitsUsageCostShared,
                                    spendAckChecked: sharedThroughputSpendAck,
                                    spendAckId: 'sharedThroughputSpendAck-v2-shared',
                                    spendAckText: sharedThroughputSpendAckText,
                                    spendAckVisible: sharedThroughputSpendAckVisible,
                                    showAsMandatory: true,
                                    infoBubbleText: ruToolTipText,
                                    throughputAutoPilotRadioId: 'newKeyspace-databaseThroughput-autoPilotRadio-v2-shared',
                                    throughputProvisionedRadioId: 'newKeyspace-databaseThroughput-manualRadio-v2-shared',
                                    isAutoPilotSelected: isSharedAutoPilotSelected,
                                    autoPilotTiersList: sharedAutoPilotTiersList,
                            costsVisible: costsVisible,
                                    selectedAutoPilotTier: selectedSharedAutoPilotTier,
                                    autoPilotUsageCost: autoPilotUsageCost,
                                    canExceedMaximumValue: canExceedMaximumValue
                                    }"
              >
              </throughput-input>
            </div>
            <!-- /ko -->
          </div>
          <div class="seconddivpadding">
            <p>
              <span class="mandatoryStar">*</span> Enter CQL command to create the table.
              <a href="https://aka.ms/cassandra-create-table" target="_blank">Learn More</a>
            </p>
            <div data-bind="text: createTableQuery" style="float: left; padding-top: 3px; padding-right: 3px;"></div>
            <input
              type="text"
              data-test="addCollection-tableId"
              aria-required="true"
              autocomplete="off"
              pattern="[^/?#\\]*[^/?# \\]"
              title="May not end with space nor contain characters '\' '/' '#' '?'"
              data-test="addCollection-tableId"
              placeholder="Enter tableId"
              size="20"
              class="textfontclr"
              data-bind="value: tableId"
              style="margin-bottom: 5px;"
            />
            <textarea
              id="editor-area"
              rows="15"
              aria-label="Table Schema"
              data-bind="value: userTableQuery"
              style="height: 125px; width: calc(100% - 80px); resize: vertical;"
            ></textarea>
          </div>
          <div class="seconddivpadding" data-bind="visible: keyspaceHasSharedOffer() && !keyspaceCreateNew()">
            <input
              type="checkbox"
              id="tableSharedThroughput"
              title="Provision dedicated throughput for this table"
              data-bind="checked: dedicateTableThroughput"
            />
            <span for="tableSharedThroughput">Provision dedicated throughput for this table</span>
            <span class="leftAlignInfoTooltip" role="tooltip" tabindex="0">
              <img class="infoImg" src="/info-bubble.svg" alt="More information" />
              <span class="tooltiptext sharedCollectionThroughputTooltipWidth"
                >You can optionally provision dedicated throughput for a table within a keyspace that has throughput
                provisioned. This dedicated throughput amount will not be shared with other tables in the keyspace and
                does not count towards the throughput you provisioned for the keyspace. This throughput amount will be
                billed in addition to the throughput amount you provisioned at the keyspace level.</span
              >
            </span>
          </div>
          <!-- 2 -->
          <!-- ko if: hasAutoPilotV2FeatureFlag && !hasAutoPilotV2FeatureFlag() -->
          <div data-bind="visible: !keyspaceHasSharedOffer() || dedicateTableThroughput()">
            <throughput-input-autopilot-v3
              params="{ 
                            testId: 'cassandraSharedThroughputValue-v3-dedicated',
                            value: throughput, 
                            minimum: minThroughputRU, 
                            maximum: maxThroughputRU, 
                            isEnabled: !keyspaceHasSharedOffer() || dedicateTableThroughput(), 
                            label: throughputRangeText,
                            ariaLabel: throughputRangeText,
                            costsVisible: costsVisible,
                            requestUnitsUsageCost: requestUnitsUsageCostDedicated,
                            spendAckChecked: throughputSpendAck,
                            spendAckId: 'throughputSpendAckCassandra-v3-dedicated',
                            spendAckText: throughputSpendAckText,
                            spendAckVisible: throughputSpendAckVisible,
                            showAsMandatory: true,
                            infoBubbleText: ruToolTipText,
                            throughputAutoPilotRadioId: 'newKeyspace-containerThroughput-autoPilotRadio-v3-dedicated',
                            throughputProvisionedRadioId: 'newKeyspace-containerThroughput-manualRadio-v3-dedicated',
                            isAutoPilotSelected: isAutoPilotSelected,
                            maxAutoPilotThroughputSet: selectedAutoPilotThroughput,
                            autoPilotUsageCost: autoPilotUsageCost,
                            canExceedMaximumValue: canExceedMaximumValue,
                            overrideWithAutoPilotSettings: false,
                        overrideWithProvisionedThroughputSettings: false
                        }"
            >
            </throughput-input-autopilot-v3>
          </div>
          <!-- /ko -->

          <!-- ko if: hasAutoPilotV2FeatureFlag && hasAutoPilotV2FeatureFlag() -->
          <div data-bind="visible: !keyspaceHasSharedOffer() || dedicateTableThroughput()">
            <throughput-input
              params="{ 
                            testId: 'cassandraSharedThroughputValue-v2-dedicated',
                            value: throughput, 
                            minimum: minThroughputRU, 
                            maximum: maxThroughputRU, 
                            isEnabled: !keyspaceHasSharedOffer() || dedicateTableThroughput(), 
                            label: throughputRangeText,
                            ariaLabel: throughputRangeText,
                            costsVisible: costsVisible,
                            requestUnitsUsageCost: requestUnitsUsageCostDedicated,
                            spendAckChecked: throughputSpendAck,
                            spendAckId: 'throughputSpendAckCassandra-v2-dedicated',
                            spendAckText: throughputSpendAckText,
                            spendAckVisible: throughputSpendAckVisible,
                            showAsMandatory: true,
                            infoBubbleText: ruToolTipText,
                            throughputAutoPilotRadioId: 'newKeyspace-containerThroughput-autoPilotRadio-v2-dedicated',
                            throughputProvisionedRadioId: 'newKeyspace-containerThroughput-manualRadio-v2-dedicated',
                            isAutoPilotSelected: isAutoPilotSelected,
                            autoPilotTiersList: autoPilotTiersList,
                            selectedAutoPilotTier: selectedAutoPilotTier,
                            autoPilotUsageCost: autoPilotUsageCost,
                            showAutoPilot: false,
                            canExceedMaximumValue: canExceedMaximumValue
                        }"
            >
            </throughput-input>
          </div>
          <!-- /ko -->
        </div>
        <div class="paneFooter">
          <div class="leftpanel-okbut">
            <input type="submit" data-test="addCollection-createCollection" value="OK" class="btncreatecoll1" />
          </div>
        </div>
      </form>
      <!-- Add Cassandra collection form - End -->
      <!-- Loader - Start -->
      <div class="dataExplorerLoaderContainer dataExplorerPaneLoaderContainer" data-bind="visible: isExecuting">
        <img class="dataExplorerLoader" src="/LoadingIndicator_3Squares.gif" alt="loading indicator" />
      </div>
      <!-- Loader - End -->
    </div>
  </div>
</div>
