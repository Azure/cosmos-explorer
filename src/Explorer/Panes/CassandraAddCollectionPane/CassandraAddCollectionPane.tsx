import { Checkbox, Dropdown, IDropdownOption, Link, Stack, Text, TextField } from "@fluentui/react";
import * as Constants from "Common/Constants";
import { getErrorMessage, getErrorStack } from "Common/ErrorHandlingUtils";
import { InfoTooltip } from "Common/Tooltip/InfoTooltip";
import * as SharedConstants from "Shared/Constants";
import { Action } from "Shared/Telemetry/TelemetryConstants";
import * as TelemetryProcessor from "Shared/Telemetry/TelemetryProcessor";
import { userContext } from "UserContext";
import { isServerlessAccount } from "Utils/CapabilityUtils";
import { ValidCosmosDbIdDescription, ValidCosmosDbIdInputPattern } from "Utils/ValidationUtils";
import { useSidePanel } from "hooks/useSidePanel";
import React, { FunctionComponent, useState } from "react";
import { ThroughputInput } from "../../Controls/ThroughputInput/ThroughputInput";
import Explorer from "../../Explorer";
import { CassandraAPIDataClient } from "../../Tables/TableDataClient";
import { useDatabases } from "../../useDatabases";
import { getTextFieldStyles } from "../PanelStyles";
import { RightPaneForm, RightPaneFormProps } from "../RightPaneForm/RightPaneForm";

export interface CassandraAddCollectionPaneProps {
  explorer: Explorer;
  cassandraApiClient: CassandraAPIDataClient;
}

export const CassandraAddCollectionPane: FunctionComponent<CassandraAddCollectionPaneProps> = ({
  explorer: container,
  cassandraApiClient,
}: CassandraAddCollectionPaneProps) => {
  let newKeySpaceThroughput: number;
  let isNewKeySpaceAutoscale: boolean;
  let tableThroughput: number;
  let isTableAutoscale: boolean;
  let isCostAcknowledged: boolean;

  const closeSidePanel = useSidePanel((state) => state.closeSidePanel);
  const [newKeyspaceId, setNewKeyspaceId] = useState<string>("");
  const [existingKeyspaceId, setExistingKeyspaceId] = useState<string>("");
  const [tableId, setTableId] = useState<string>("");
  const [userTableQuery, setUserTableQuery] = useState<string>(
    "(userid int, name text, email text, PRIMARY KEY (userid))",
  );
  const [isKeyspaceShared, setIsKeyspaceShared] = useState<boolean>(false);
  const [keyspaceCreateNew, setKeyspaceCreateNew] = useState<boolean>(true);
  const [dedicateTableThroughput, setDedicateTableThroughput] = useState<boolean>(false);
  const [isExecuting, setIsExecuting] = useState<boolean>();
  const [formError, setFormError] = useState<string>("");
  const [isThroughputCapExceeded, setIsThroughputCapExceeded] = useState<boolean>(false);
  const isFreeTierAccount: boolean = userContext.databaseAccount?.properties?.enableFreeTier;

  const addCollectionPaneOpenMessage = {
    collection: {
      id: tableId,
      storage: Constants.BackendDefaults.multiPartitionStorageInGb,
      offerThroughput: newKeySpaceThroughput || tableThroughput,
      partitionKey: "",
      databaseId: keyspaceCreateNew ? newKeyspaceId : existingKeyspaceId,
    },
    subscriptionType: userContext.subscriptionType,
    subscriptionQuotaId: userContext.quotaId,
    defaultsCheck: {
      storage: "u",
      throughput: newKeySpaceThroughput || tableThroughput,
    },
    dataExplorerArea: Constants.Areas.ContextualPane,
  };

  const onSubmit = async () => {
    const throughput = keyspaceCreateNew ? newKeySpaceThroughput : tableThroughput;
    const keyspaceId = keyspaceCreateNew ? newKeyspaceId : existingKeyspaceId;

    if (throughput > SharedConstants.CollectionCreation.DefaultCollectionRUs100K && !isCostAcknowledged) {
      const errorMessage =
        isNewKeySpaceAutoscale || isTableAutoscale
          ? "Please acknowledge the estimated monthly spend."
          : "Please acknowledge the estimated daily spend.";
      setFormError(errorMessage);
      return;
    }

    setIsExecuting(true);
    const autoPilotCommand = `cosmosdb_autoscale_max_throughput`;
    const createKeyspaceQueryPrefix = `CREATE KEYSPACE ${keyspaceId.trim()} WITH REPLICATION = { 'class' : 'SimpleStrategy', 'replication_factor' : 3 }`;
    const createKeyspaceQuery: string = isKeyspaceShared
      ? isNewKeySpaceAutoscale
        ? `${createKeyspaceQueryPrefix} AND ${autoPilotCommand}=${newKeySpaceThroughput};`
        : `${createKeyspaceQueryPrefix} AND cosmosdb_provisioned_throughput=${newKeySpaceThroughput};`
      : `${createKeyspaceQueryPrefix};`;
    let tableQuery: string;
    const createTableQueryPrefix = `CREATE TABLE ${keyspaceId}.${tableId.trim()} ${userTableQuery}`;

    if (tableThroughput) {
      if (isTableAutoscale) {
        tableQuery = `${createTableQueryPrefix} WITH ${autoPilotCommand}=${tableThroughput};`;
      } else {
        tableQuery = `${createTableQueryPrefix} WITH cosmosdb_provisioned_throughput=${tableThroughput};`;
      }
    } else {
      tableQuery = `${createTableQueryPrefix};`;
    }

    const addCollectionPaneStartMessage = {
      ...addCollectionPaneOpenMessage,
      collection: {
        ...addCollectionPaneOpenMessage.collection,
        hasDedicatedThroughput: dedicateTableThroughput,
      },
      isKeyspaceShared,
      keyspaceCreateNew,
      createKeyspaceQuery,
      createTableQuery: tableQuery,
    };

    const startKey: number = TelemetryProcessor.traceStart(Action.CreateCollection, addCollectionPaneStartMessage);
    try {
      if (keyspaceCreateNew) {
        await cassandraApiClient.createTableAndKeyspace(
          userContext?.databaseAccount?.properties?.cassandraEndpoint,
          userContext?.databaseAccount?.id,
          container,
          tableQuery,
          createKeyspaceQuery,
        );
      } else {
        await cassandraApiClient.createTableAndKeyspace(
          userContext?.databaseAccount?.properties?.cassandraEndpoint,
          userContext?.databaseAccount?.id,
          container,
          tableQuery,
        );
      }
      container.refreshAllDatabases();
      setIsExecuting(false);
      closeSidePanel();

      TelemetryProcessor.traceSuccess(Action.CreateCollection, addCollectionPaneStartMessage, startKey);
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      setFormError(errorMessage);
      setIsExecuting(false);
      const addCollectionPaneFailedMessage = {
        ...addCollectionPaneStartMessage,
        error: errorMessage,
        errorStack: getErrorStack(error),
      };
      TelemetryProcessor.traceFailure(Action.CreateCollection, addCollectionPaneFailedMessage, startKey);
    }
  };

  const props: RightPaneFormProps = {
    formError,
    isExecuting,
    submitButtonText: "OK",
    isSubmitButtonDisabled: isThroughputCapExceeded,
    onSubmit,
  };

  return (
    <RightPaneForm {...props}>
      <div className="panelMainContent">
        <Stack>
          <Stack horizontal>
            <span className="mandatoryStar">*&nbsp;</span>
            <Text className="panelTextBold" variant="small">
              Keyspace name <InfoTooltip>Select an existing keyspace or enter a new keyspace id.</InfoTooltip>
            </Text>
          </Stack>

          <Stack horizontal verticalAlign="center">
            <input
              className="panelRadioBtn"
              aria-label="Create new keyspace"
              checked={keyspaceCreateNew}
              type="radio"
              role="radio"
              tabIndex={0}
              onChange={() => {
                setKeyspaceCreateNew(true);
                setIsKeyspaceShared(false);
                setExistingKeyspaceId("");
              }}
            />
            <span className="panelRadioBtnLabel">Create new</span>

            <input
              className="panelRadioBtn"
              aria-label="Use existing keyspace"
              checked={!keyspaceCreateNew}
              type="radio"
              role="radio"
              tabIndex={0}
              onChange={() => {
                setKeyspaceCreateNew(false);
                setIsKeyspaceShared(false);
              }}
            />
            <span className="panelRadioBtnLabel">Use existing</span>
          </Stack>

          {keyspaceCreateNew && (
            <Stack className="panelGroupSpacing">
              <TextField
                aria-required="true"
                required={true}
                autoComplete="off"
                styles={getTextFieldStyles()}
                pattern={ValidCosmosDbIdInputPattern.source}
                title={ValidCosmosDbIdDescription}
                placeholder="Type a new keyspace id"
                size={40}
                value={newKeyspaceId}
                onChange={(e, newValue) => setNewKeyspaceId(newValue)}
                ariaLabel="Keyspace id"
                autoFocus
              />

              {!isServerlessAccount() && (
                <Stack horizontal>
                  <Checkbox
                    label="Provision shared throughput"
                    checked={isKeyspaceShared}
                    styles={{
                      text: { fontSize: 12 },
                      checkbox: { width: 12, height: 12 },
                      label: { padding: 0, alignItems: "center" },
                    }}
                    onChange={(ev: React.FormEvent<HTMLElement>, isChecked: boolean) => setIsKeyspaceShared(isChecked)}
                  />
                  <InfoTooltip>
                    Provisioned throughput at the keyspace level will be shared across unlimited number of tables within
                    the keyspace
                  </InfoTooltip>
                </Stack>
              )}
            </Stack>
          )}

          {!keyspaceCreateNew && (
            <Dropdown
              ariaLabel="Choose existing keyspace id"
              styles={{ root: { width: 300 }, title: { fontSize: 12 }, dropdownItem: { fontSize: 12 } }}
              placeholder="Choose existing keyspace id"
              defaultSelectedKey={existingKeyspaceId}
              options={useDatabases.getState().databases?.map((keyspace) => ({
                key: keyspace.id(),
                text: keyspace.id(),
                data: {
                  isShared: !!keyspace.offer(),
                },
              }))}
              onChange={(event: React.FormEvent<HTMLDivElement>, option: IDropdownOption) => {
                setExistingKeyspaceId(option.key as string);
                setIsKeyspaceShared(option.data.isShared);
              }}
              responsiveMode={999}
            />
          )}

          {!isServerlessAccount() && keyspaceCreateNew && isKeyspaceShared && (
            <ThroughputInput
              showFreeTierExceedThroughputTooltip={
                isFreeTierAccount && !useDatabases.getState().isFirstResourceCreated()
              }
              isDatabase
              isSharded
              isFreeTier={isFreeTierAccount}
              setThroughputValue={(throughput: number) => (newKeySpaceThroughput = throughput)}
              setIsAutoscale={(isAutoscale: boolean) => (isNewKeySpaceAutoscale = isAutoscale)}
              setIsThroughputCapExceeded={(isCapExceeded: boolean) => setIsThroughputCapExceeded(isCapExceeded)}
              onCostAcknowledgeChange={(isAcknowledged: boolean) => (isCostAcknowledged = isAcknowledged)}
            />
          )}
        </Stack>

        <Stack>
          <Stack horizontal>
            <span className="mandatoryStar">*&nbsp;</span>
            <Text className="panelTextBold" variant="small">
              Enter CQL command to create the table.{" "}
              <Link className="underlinedLink" href="https://aka.ms/cassandra-create-table" target="_blank">
                Learn More
              </Link>
            </Text>
          </Stack>

          <Stack horizontal verticalAlign="center">
            <Text variant="small" style={{ marginRight: 4 }}>
              {`CREATE TABLE ${keyspaceCreateNew ? newKeyspaceId : existingKeyspaceId}.`}
            </Text>
            <TextField
              underlined
              styles={getTextFieldStyles({ fontSize: 12, width: 150 })}
              aria-required="true"
              required={true}
              ariaLabel="addCollection-table Id Create table"
              autoComplete="off"
              pattern={ValidCosmosDbIdInputPattern.source}
              title={ValidCosmosDbIdDescription}
              placeholder="Enter table Id"
              size={20}
              value={tableId}
              onChange={(e, newValue) => setTableId(newValue)}
            />
          </Stack>

          <TextField
            styles={getTextFieldStyles()}
            multiline
            id="editor-area"
            rows={5}
            ariaLabel="Table schema"
            value={userTableQuery}
            onChange={(e, newValue) => setUserTableQuery(newValue)}
          />
        </Stack>

        {!isServerlessAccount() && isKeyspaceShared && !keyspaceCreateNew && (
          <Stack horizontal verticalAlign="center">
            <input
              type="checkbox"
              id="tableSharedThroughput"
              title="Provision dedicated throughput for this table"
              checked={dedicateTableThroughput}
              onChange={(e) => setDedicateTableThroughput(e.target.checked)}
            />
            <span>Provision dedicated throughput for this table</span>
            <InfoTooltip>
              You can optionally provision dedicated throughput for a table within a keyspace that has throughput
              provisioned. This dedicated throughput amount will not be shared with other tables in the keyspace and
              does not count towards the throughput you provisioned for the keyspace. This throughput amount will be
              billed in addition to the throughput amount you provisioned at the keyspace level.
            </InfoTooltip>
          </Stack>
        )}
        {!isServerlessAccount() && (!isKeyspaceShared || dedicateTableThroughput) && (
          <ThroughputInput
            showFreeTierExceedThroughputTooltip={isFreeTierAccount && !useDatabases.getState().isFirstResourceCreated()}
            isDatabase={false}
            isSharded
            isFreeTier={isFreeTierAccount}
            setThroughputValue={(throughput: number) => (tableThroughput = throughput)}
            setIsAutoscale={(isAutoscale: boolean) => (isTableAutoscale = isAutoscale)}
            setIsThroughputCapExceeded={(isCapExceeded: boolean) => setIsThroughputCapExceeded(isCapExceeded)}
            onCostAcknowledgeChange={(isAcknowledged: boolean) => (isCostAcknowledged = isAcknowledged)}
          />
        )}
      </div>
    </RightPaneForm>
  );
};
