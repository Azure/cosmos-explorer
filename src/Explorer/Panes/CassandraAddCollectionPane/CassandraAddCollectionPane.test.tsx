import { fireEvent, render, screen } from "@testing-library/react";
import React from "react";
import Explorer from "../../Explorer";
import { CassandraAPIDataClient } from "../../Tables/TableDataClient";
import { CassandraAddCollectionPane } from "./CassandraAddCollectionPane";

describe("Cassandra add collection pane test", () => {
  const props = {
    explorer: new Explorer(),
    closePanel: (): void => undefined,
    cassandraApiClient: new CassandraAPIDataClient(),
  };

  beforeEach(() => render(<CassandraAddCollectionPane {...props} />));

  it("should render default properly", () => {
    expect(screen.getByRole("radio", { name: "Create new keyspace", checked: true })).toBeDefined();
    expect(screen.getByRole("checkbox", { name: "Provision shared throughput", checked: false })).toBeDefined();
  });

  it("click on use existing", () => {
    fireEvent.click(screen.getByRole("radio", { name: "Use existing keyspace" }));
    expect(screen.getByRole("combobox", { name: "Choose existing keyspace id" })).toBeDefined();
  });

  it("enter Keyspace name ", () => {
    fireEvent.change(screen.getByRole("textbox", { name: "Keyspace id" }), { target: { value: "table1" } });
    expect(screen.getByText("CREATE TABLE table1.")).toBeDefined();
  });
});
