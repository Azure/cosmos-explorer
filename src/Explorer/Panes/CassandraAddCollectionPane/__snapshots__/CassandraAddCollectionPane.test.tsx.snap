// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CassandraAddCollectionPane  Pane should render Default properly 1`] = `
<RightPaneForm
  formError=""
  onSubmit={[Function]}
  submitButtonText="Apply"
>
  <div
    className="paneMainContent"
  >
    <div
      className="seconddivpadding"
    >
      <p>
        <StyledLabelBase
          required={true}
        >
          Keyspace name 
          <InfoTooltip>
            Select an existing keyspace or enter a new keyspace id.
          </InfoTooltip>
        </StyledLabelBase>
      </p>
      <Stack
        horizontal={true}
        verticalAlign="center"
      >
        <input
          aria-label="Create new keyspace"
          checked={true}
          className="throughputInputRadioBtn"
          onChange={[Function]}
          role="radio"
          tabIndex={0}
          type="radio"
        />
        <span
          className="throughputInputRadioBtnLabel"
        >
          Create new
        </span>
        <input
          aria-label="Use existing keyspace"
          checked={false}
          className="throughputInputRadioBtn"
          onChange={[Function]}
          role="radio"
          tabIndex={0}
          type="radio"
        />
        <span
          className="throughputInputRadioBtnLabel"
        >
          Use existing
        </span>
      </Stack>
      <StyledTextFieldBase
        aria-required="true"
        ariaLabel="Keyspace id"
        autoComplete="off"
        autoFocus={true}
        data-test="addCollection-keyspaceId"
        list=""
        onChange={[Function]}
        pattern="[^/?#\\\\\\\\]*[^/?# \\\\\\\\]"
        placeholder="Type a new keyspace id"
        size={40}
        title="May not end with space nor contain characters '\\\\' '/' '#' '?'"
        value=""
      />
      <datalist
        id="keyspacesList"
      />
      <div
        className="databaseProvision"
      >
        <input
          checked={false}
          id="keyspaceSharedThroughput"
          onChange={[Function]}
          tabIndex={0}
          title="Provision shared throughput"
          type="checkbox"
        />
        <span
          aria-label="Provision keyspace throughput"
          className="databaseProvisionText"
        >
          Provision keyspace throughput
        </span>
        <InfoTooltip>
          Provisioned throughput at the keyspace level will be shared across unlimited number of tables within the keyspace
        </InfoTooltip>
      </div>
    </div>
    <div
      className="seconddivpadding"
    >
      <p>
        <StyledLabelBase
          required={true}
        >
          Enter CQL command to create the table.
          <a
            href="https://aka.ms/cassandra-create-table"
            rel="noreferrer"
            target="_blank"
          >
            Learn More
          </a>
        </StyledLabelBase>
      </p>
      <div
        aria-label="CREATE TABLE "
        style={
          Object {
            "float": "left",
            "paddingRight": "3px",
            "paddingTop": "3px",
          }
        }
      >
        CREATE TABLE 
      </div>
      <StyledTextFieldBase
        aria-required="true"
        ariaLabel="addCollection-tableId"
        autoComplete="off"
        className="textfontclr"
        onChange={[Function]}
        pattern="[^/?#\\\\\\\\]*[^/?# \\\\\\\\]"
        placeholder="Enter tableId"
        size={20}
        style={
          Object {
            "marginBottom": "5px",
          }
        }
        title="May not end with space nor contain characters '\\\\' '/' '#' '?'"
        value=""
      />
      <StyledTextFieldBase
        aria-label="Table Schema"
        id="editor-area"
        multiline={true}
        onChange={[Function]}
        rows={5}
        value="(userid int, name text, email text, PRIMARY KEY (userid))"
      />
    </div>
    <div>
      <ThroughputInput
        isDatabase={false}
        isSharded={false}
        onCostAcknowledgeChange={[Function]}
        setIsAutoscale={[Function]}
        setThroughputValue={[Function]}
      />
    </div>
  </div>
</RightPaneForm>
`;
